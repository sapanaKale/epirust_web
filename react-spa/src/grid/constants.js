export const BaseColors = {
    YELLOW: "#fcce44",
    LIGHT_YELLOW: "#fff0c2",
    PURPLE: "#3498db",
    BLUE: "#007bff",
    ORANGE: "#ffa500",
    RED: "#e74c3c",
    GRAY_LIGHTER: "#f1f1f1",
    PINK: "#e83e8c",
    LIGHT_PINK: "#f9d1e4",
    GRAY: "#ccc",
    BLACK: "black",
    GREEN: "#28a745",
    WHITE: "#fff",
    LIGHT_ORANGE: "#fad5b5",
    LIGHT_BLUE: "#b9dafd"
};

export const AreaColors = {
    HOUSING: BaseColors.LIGHT_YELLOW,
    WORK: BaseColors.LIGHT_PINK,
    TRANSPORT: BaseColors.LIGHT_ORANGE,
    HOSPITAL: BaseColors.LIGHT_BLUE,
    OTHER: BaseColors.GRAY
};

export const LandmarkColors = {
    HOUSES: BaseColors.GRAY,
    OFFICES: BaseColors.GRAY
};

export const AgentStateToColor = {
    's': BaseColors.BLUE,
    'e': BaseColors.ORANGE,
    'i': BaseColors.RED,
    'r': BaseColors.GREEN,
    'd': BaseColors.BLACK
};

export const AgentStateMapper = {
    's': 'susceptible',
    'e': 'exposed',
    'i': 'infected',
    'r': 'recovered',
    'd': 'deceased'
};

export const Interventions = Object.freeze({
    LOCKDOWN: 'lockdown',
    BUILD_NEW_HOSPITAL: 'build_new_hospital',
    VACCINATION: 'vaccination',
    status: {
        LOCKDOWN_START: 'locked_down',
    }
});
