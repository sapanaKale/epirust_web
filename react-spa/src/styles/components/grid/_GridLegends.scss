.legends-wrap {
    border: 2px solid powderblue;
    position: absolute;
    top: 240px;
    right: 32px;

    .legends {
        padding: 5px;
        list-style: none;

        .legend-item {
            margin-bottom: 5px;
        }

        .swatch {
            margin-right: 4px;
            display: inline-block;
        }
        .text {
            vertical-align: text-bottom;
            font-size: 12px;
        }

        .count {
            padding-right: 4px;
        }


        &.areas {
            .swatch {
                padding: 8px 15px;
            }
        }

        &.agents {
            .swatch {
                padding: 5px;
                border-radius: 50%;
            }
        }
    }
}
