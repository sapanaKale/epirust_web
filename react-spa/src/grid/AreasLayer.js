/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
import React, { useContext, useRef, useState, useEffect } from 'react';
import { GridContext } from './index'
import { AreaColors } from './constants';

export default function AreasLayer({ areaDimensions }) {

    const { cellDimension, lineWidth, canvasDimension, size } = useContext(GridContext);

    const areasLayerCanvas = useRef(null);
    const [areasCanvasContext, setAreasCanvasContext] = useState(null);

    useEffect(() => {
        if (!areaDimensions)
            return

        if (!areaDimensions || !areasLayerCanvas)
            return

        setAreasCanvasContext(areasLayerCanvas.current.getContext("2d"));

    }, [areasLayerCanvas, areaDimensions])

    useEffect(() => {
        if (!areaDimensions || !areasCanvasContext)
            return

        function updateAreaColor(areaDimensions, x, y) {
            let area;
            for (var i = 0; i < areaDimensions.length; i++) {
                if (isWithinArea(areaDimensions[i], x, y)) {
                    area = areaDimensions[i];
                    break;
                }
            }

            const color = area ? area.color : AreaColors.OTHER;

            if (areasCanvasContext.fillStyle !== color) {
                areasCanvasContext.fillStyle = color;
            }

            // areasCanvasContext.fillStyle = "white";
        }

        for (let x = 0; x < size; x++) {
            for (let y = 0; y < size; y++) {
                updateAreaColor(areaDimensions, x, y);
                areasCanvasContext.fillRect((x * cellDimension) + lineWidth / 2, (y * cellDimension) + lineWidth / 2, cellDimension, cellDimension)
            }
        }
    }, [areasCanvasContext, size, cellDimension, lineWidth, areaDimensions])

    return (
        <canvas ref={areasLayerCanvas} data-testid="grid-canvas" id="grid-canvas" width={canvasDimension} height={canvasDimension} style={{ position: "absolute", zIndex: 1 }} />
    )
}

function isWithinArea(area, currentX, currentY) {

    return (currentX >= area.start_offset.x && currentY >= area.start_offset.y
        && currentX <= area.end_offset.x && currentY <= area.end_offset.y);
}
