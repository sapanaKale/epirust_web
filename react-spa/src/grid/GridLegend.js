/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from 'react';
import { AgentStateToColor, AreaColors } from './constants';
import PropTypes from 'prop-types'

export default function GridLegend({ susceptible, exposed, infected, recovered, deceased }) {

    const areasLegends = [
        { backgroundColor: AreaColors.HOUSING, text: "Housing" },
        { backgroundColor: AreaColors.WORK, text: "Work" },
        { backgroundColor: AreaColors.TRANSPORT, text: "Transport" },
        { backgroundColor: AreaColors.HOSPITAL, text: "Hospital" },
        { backgroundColor: AreaColors.OTHER, text: "Other" }
    ];

    const agentsLegends = [
        { backgroundColor: AgentStateToColor.s, text: "Susceptible", count: susceptible },
        { backgroundColor: AgentStateToColor.e, text: "Exposed", count: exposed },
        { backgroundColor: AgentStateToColor.i, text: "Infected", count: infected },
        { backgroundColor: AgentStateToColor.r, text: "Recovered", count: recovered },
        { backgroundColor: AgentStateToColor.d, text: "Deceased", count: deceased }
    ];
    return (
        <div className="legends-wrap px-2 py-2">
            <h6>Areas</h6>
            <ul className="legends areas">{
                areasLegends.map(({ backgroundColor, text }) => (
                    <li key={text}>
                        <div className="legend-item">
                            <span className="swatch" style={{ backgroundColor }} />
                            <span className="text">{text}</span>
                        </div>
                    </li>
                ))}
            </ul>

            <h6>Agent States</h6>
            <ul className="legends agents">{
                agentsLegends.map(({ backgroundColor, text, count }) => (
                    <li key={text}>
                        <div className="legend-item">
                            <span className="swatch" style={{ backgroundColor }} />
                            <span className="text count">{count}</span>
                            <span className="text">{text}</span>
                        </div>
                    </li>
                ))}
            </ul>
        </div>
    )
}

GridLegend.propTypes = {
    susceptible: PropTypes.number,
    exposed: PropTypes.number,
    infected: PropTypes.number,
    recovered: PropTypes.number,
    deceased: PropTypes.number
};
