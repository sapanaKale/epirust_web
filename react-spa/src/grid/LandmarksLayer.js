/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
import React, { useContext, useRef, useState, useEffect } from 'react';
import { GridContext } from './index'
import { LandmarkColors } from './constants';

export default function LandmarksLayer({ landmarksDimensions }) {
    const { cellDimension, lineWidth, canvasDimension, size } = useContext(GridContext);

    const landmarksLayerCanvas = useRef(null);
    const [landmarksCanvasContext, setLandmarksCanvasContext] = useState(null);

    useEffect(() => {
        if (!landmarksDimensions)
            return

        if (!landmarksLayerCanvas)
            return

        setLandmarksCanvasContext(landmarksLayerCanvas.current.getContext("2d"));

    }, [landmarksLayerCanvas, landmarksDimensions])

    useEffect(() => {
        if (!landmarksDimensions)
            return

        if (!landmarksCanvasContext)
            return

        landmarksCanvasContext.lineWidth = lineWidth;

        [
            { landmark: landmarksDimensions.housesDimensions, color: LandmarkColors.HOUSES },
            { landmark: landmarksDimensions.officesDimensions, color: LandmarkColors.OFFICES }
        ].forEach((landmarkObject) => {
            landmarksCanvasContext.strokeStyle = landmarkObject.color

            landmarkObject.landmark.forEach((element) => {

                const startX = element.start_offset.x;
                const startY = element.start_offset.y;

                const width = 1 + element.end_offset.x - startX
                const height = 1 + element.end_offset.y - startY

                /**
                 *   ____________
                 *  | 0,0 | 0, 1|
                 *  |_____|_____|
                 *  | 0,0 | 1, 1|
                 *  |_____|_____|
                 * e.g. for a landmark, start_offset {x: 0, y: 0} & end_offset {x: 1, y: 1}
                 * Hence, width and height of the landmark is 2 cells.
                 */

                landmarksCanvasContext.strokeRect((startX * cellDimension) + lineWidth / 2, (startY * cellDimension) + lineWidth / 2, width * cellDimension, height * cellDimension);
            });
        });
    }, [landmarksCanvasContext, cellDimension, lineWidth, landmarksDimensions])

    return (
        <canvas ref={landmarksLayerCanvas} data-testid="grid-canvas-landmarks" id="grid-canvas-landmarks" width={canvasDimension} height={canvasDimension} style={{ position: "absolute", zIndex: 3 }} />
    )
}