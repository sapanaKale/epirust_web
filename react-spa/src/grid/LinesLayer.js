/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, { useContext, useRef, useState, useEffect } from 'react';
import { GridContext } from './index'
import { BaseColors } from './constants';
export default function LinesLayer() {

    const { cellDimension, lineWidth, canvasDimension, size } = useContext(GridContext);

    const linesLayerCanvas = useRef(null);
    const [lineCanvasContext, setLineCanvasContext] = useState(null);

    useEffect(() => {
        if (!linesLayerCanvas)
            return

        setLineCanvasContext(linesLayerCanvas.current.getContext("2d"));

    }, [linesLayerCanvas])

    useEffect(() => {
        if (!lineCanvasContext || lineWidth === 0)
            return

        lineCanvasContext.strokeStyle = BaseColors.GRAY_LIGHTER;
        lineCanvasContext.lineWidth = lineWidth;

        for (let x = 0; x < size; x++) {
            for (let y = 0; y < size; y++) {
                lineCanvasContext.strokeRect((x * cellDimension) + lineWidth / 2, (y * cellDimension) + lineWidth / 2, cellDimension, cellDimension);
            }
        }

    }, [lineCanvasContext, size, cellDimension, lineWidth])

    return (
        <canvas ref={linesLayerCanvas} data-testid="grid-canvas-bg" id="grid-canvas-bg" width={canvasDimension} height={canvasDimension} style={{ position: "absolute", zIndex: 2 }} />
    )
}