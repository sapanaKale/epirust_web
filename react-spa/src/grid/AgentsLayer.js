/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
import React, { useContext, useEffect, useRef, useState } from 'react';
import { GridContext } from './index'
import { AgentStateMapper, AgentStateToColor } from './constants';
import GridLegend from "./GridLegend";
import {faPauseCircle, faPlayCircle, faRedo, faStepForward, faStepBackward} from "@fortawesome/free-solid-svg-icons";
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";

export default function AgentPositionsWrapper({ agentPositions, simulationEnded }) {
    const [displayIndexIncrementPaused, setDisplayIndexIncrementPaused] = useState(true);
    const [currentDisplayIndex, setCurrentDisplayIndex] = useState(0);
    const [intervalId, setIntervalId] = useState(null);
    const [clickedPause, setClickedPause] = useState(false);

    const stopIncrement = () => {
        clearInterval(intervalId);
        setIntervalId(null);
    };

    const displayedAll = () => {
        return simulationEnded && (currentDisplayIndex >= agentPositions.length - 1)
    };

    //displayed all the data
    useEffect(() => {
        if (displayedAll()) {
            stopIncrement()
        }
    }, [currentDisplayIndex, agentPositions]);

    //pause
    useEffect(() => {
        if (displayIndexIncrementPaused && agentPositions) {
            stopIncrement();
        }
    }, [displayIndexIncrementPaused, agentPositions]);

    //race condition
    useEffect(() => {
        if (!agentPositions || simulationEnded) {
            return
        }
        if (currentDisplayIndex >= agentPositions.length - 1) {
            setDisplayIndexIncrementPaused(true);
            return
        }

        if (displayIndexIncrementPaused && !clickedPause) {
            setDisplayIndexIncrementPaused(false)
        }

    }, [currentDisplayIndex, agentPositions, simulationEnded, clickedPause]);

    //count++
    useEffect(() => {
        if (displayIndexIncrementPaused)
            return;

        if (!agentPositions || intervalId)
            return;

        const interval = setInterval(() => setCurrentDisplayIndex(index => index + 1), 100);
        console.log("interval started", interval);
        setIntervalId(interval);

        return () => clearInterval(intervalId)
    }, [displayIndexIncrementPaused, agentPositions]);

    const handleResume = () => {
        setClickedPause(false);
        setDisplayIndexIncrementPaused(false)
    };

    const handlePause = () => {
        setClickedPause(true);
        setDisplayIndexIncrementPaused(true)
    };

    const handleReset = () => {
        setCurrentDisplayIndex(0);
        setClickedPause(true);
        setDisplayIndexIncrementPaused(true)
    };

    const positionsToDisplay = agentPositions
        ? agentPositions[currentDisplayIndex]
        : [];

    const getCounts = (citizens) => {
        const counts = {
            'susceptible': 0,
            'exposed': 0,
            'infected': 0,
            'recovered': 0,
            'deceased': 0
        };
        citizens && citizens.forEach(a => counts[AgentStateMapper[a.state]] += 1);
        return counts

    };

    const handleAddHour = () => {
        if (currentDisplayIndex + 1 >= agentPositions.length)
            return

        setCurrentDisplayIndex(idx => idx + 1)
    }

    const handleSubtractHour = () => {
        if (currentDisplayIndex <= 0)
            return

        setCurrentDisplayIndex(idx => idx - 1)
    }

    const dayAndHour = (index) => {
        let day = Math.floor((index + 1) / 24);
        let hour = (index + 1) % 24;
        return `Day ${day}, hour: ${hour}`;
    }

    return (
        <div style={{ position: "relative" }}>
            <div className="card p-4 grid-control" style={{ position: "absolute", zIndex: 5, right: 0, width: 190 }}>
                <h4 style={{ textAlign: 'center' }} className="pb-2" data-testid="counter">{`${currentDisplayIndex + 1}/${agentPositions ? agentPositions.length : 0} hrs`}</h4>

                <div style={{ textAlign: 'center' }} className="pb-2" >{ dayAndHour(currentDisplayIndex) }</div>

                <div className="btn-group mb-2 justify-content-center" role="group" aria-label="Buttons to pause/play and reset">
                    {clickedPause
                        ? <FontAwesomeIcon onClick={handleResume} data-testid="resume" disabled={displayedAll()} icon={faPlayCircle} className="icon resume"/>
                        : <FontAwesomeIcon data-testid="pause" icon={faPauseCircle} onClick={handlePause} disabled={displayedAll()} className="icon pause"/>
                    }

                    <FontAwesomeIcon onClick={handleReset} data-testid="reset" icon={faRedo} className="icon reset"/>
                </div>

                {(clickedPause || displayedAll()) && (
                    <div data-testid="hour-step-controls" className="btn-group my-0 mx-auto hour-step-control" role="group" aria-label="Buttons to increase/decrease current displayed hour">
                        <FontAwesomeIcon icon={faStepBackward} data-testid="decrement-hour" name="decrement hour" onClick={handleSubtractHour} className="icon"/>
                        <FontAwesomeIcon icon={faStepForward} data-testid="increment-hour" name="increment hour" onClick={handleAddHour} className="icon" />
                    </div>
                )}

            </div>
            <AgentsLayer agentPositionsPerHour={positionsToDisplay} />

            <GridLegend {...getCounts(positionsToDisplay)} />
        </div>
    )
}

function AgentsLayer({ agentPositionsPerHour }) {
    const { cellDimension, lineWidth, canvasDimension } = useContext(GridContext);

    const agentsLayerCanvas = useRef(null);
    const [agentsCanvasContext, setAgentsCanvasContext] = useState(null);

    useEffect(() => {
        if (!agentsLayerCanvas)
            return;

        setAgentsCanvasContext(agentsLayerCanvas.current.getContext("2d"));

    }, [agentsLayerCanvas]);

    useEffect(() => {
        if (!agentsCanvasContext || !agentPositionsPerHour)
            return;

        agentsCanvasContext.clearRect(0, 0, canvasDimension, canvasDimension);

        agentPositionsPerHour.forEach((agent) => {
            const { x, y } = agent.location;

            agentsCanvasContext.fillStyle = AgentStateToColor[agent.state];

            agentsCanvasContext.beginPath();
            const agentCircleRadius = Math.floor(cellDimension / 2);
            const startAngle = 0, endAngle = 2 * Math.PI;
            agentsCanvasContext.arc(calculateCoordinate(x), calculateCoordinate(y), agentCircleRadius, startAngle, endAngle);
            agentsCanvasContext.fill();
        })
    }, [agentsCanvasContext, agentPositionsPerHour]);

    function calculateCoordinate(x) {
        return Math.floor(x * cellDimension + (0.5 * cellDimension) + lineWidth)
    }

    return (
        <canvas ref={agentsLayerCanvas} data-testid="grid-canvas-agents" id="grid-canvas-agents" width={canvasDimension}
            height={canvasDimension} style={{ position: "absolute", zIndex: 4 }} />
    )
}
