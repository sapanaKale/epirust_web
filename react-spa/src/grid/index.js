/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, {useEffect, useState} from 'react';

import LandmarksLayer from './LandmarksLayer';
import LinesLayer from './LinesLayer';
import AreasLayer from './AreasLayer';
import AgentsLayer from './AgentsLayer';
import {AreaColors} from './constants';
import io from 'socket.io-client'
import config from "../config";
import Loader from '../common/Loader';
import PropTypes from "prop-types";

export const GridContext = React.createContext(null);

export default function GridPage({jobId}) {
    const [socket, setSocket] = useState(null);
    const [socketDataExhausted, setSocketDataExhausted] = useState(false);

    //default values?
    const [areaDimensions, setAreaDimensions] = useState(null);
    const [landmarksDimensions, setLandmarksDimensions] = useState(null);
    const [agentPositions, setAgentPositions] = useState(null);

    const [gridContextData, setGridContextData] = useState(null);

    useEffect(() => {
        console.log("started socket");
        const socketInstance = io(`${config.API_HOST}/grid-updates`);
        setSocket(socketInstance);

        return () => {
            socketInstance.close();
        }
    }, []);

    //reading socket data
    useEffect(() => {
        if (!socket)
            return;

        socket.emit('get', {jobId});

        socket.on('gridData', function (message) {
            if ("simulation_ended" in message) {
                socket.close();
                setSocketDataExhausted(true);
                return
            }

            if ('grid_size' in message) {
                const { housing_area, work_area, transport_area, hospital_area, grid_size } = message;
                const areaDimensions = [
                    { ...housing_area, color: AreaColors.HOUSING },
                    { ...work_area, color: AreaColors.WORK },
                    { ...transport_area, color: AreaColors.TRANSPORT },
                    { ...hospital_area, color: AreaColors.HOSPITAL }
                ];

                const housesDimensions = message.houses,
                    officesDimensions = message.offices;

                const cellDimension = Math.floor((window.innerHeight - 150) / grid_size),
                    lineWidth = Math.floor(cellDimension / 4) < 1 ? 0 : Math.floor(cellDimension / 4),
                    canvasDimension = (grid_size * cellDimension) + lineWidth;

                setGridContextData({
                    cellDimension: cellDimension,
                    lineWidth: lineWidth,
                    canvasDimension: canvasDimension,
                    size: grid_size
                });

                setAreaDimensions(areaDimensions);
                setLandmarksDimensions({ housesDimensions, officesDimensions });
                return
            }

            if ('citizen_states' in message) {

                setAgentPositions(pos => {
                    if (!pos)
                        return [message.citizen_states];

                    return [...pos, message.citizen_states]
                })
            }

        });
    }, [socket, jobId]);

    if (!gridContextData)
        return (
          <div className="grid-wrap">
              <Loader />
          </div>
        )

    return (
        <div className="grid-wrap">
            <GridContext.Provider value={gridContextData}>
                <div style={{ position: "relative" }}>
                    <AreasLayer areaDimensions={areaDimensions} />
                    <LinesLayer />
                    <LandmarksLayer landmarksDimensions={landmarksDimensions} />
                    {<AgentsLayer agentPositions={agentPositions} simulationEnded={socketDataExhausted} />}
                </div >
            </GridContext.Provider>
        </div>
    )
}

GridPage.propTypes = {
    jobId: PropTypes.string.isRequired
}