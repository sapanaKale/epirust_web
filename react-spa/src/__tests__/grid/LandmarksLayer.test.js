/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from 'react'
import { render } from '@testing-library/react'
import LandmarksLayer from '../../grid/LandmarksLayer'
import { GridContext } from '../../grid/index'

test('should render LandmarksLayer component', () => {
    const cellDimension = 2
    const lineWidth = 1
    const canvasDimension = 20
    const size = 10

    const landmarksDimensions = {
        "housesDimensions": [
            {
                "start_offset": {
                    "x": 0,
                    "y": 0
                },
                "end_offset": {
                    "x": 2,
                    "y": 2
                },
            },
            {
                "start_offset": {
                    "x": 2,
                    "y": 0
                },
                "end_offset": {
                    "x": 4,
                    "y": 2
                },
            }],
        "officesDimensions": [
            {
                "start_offset": {
                    "x": 5,
                    "y": 0
                },
                "end_offset": {
                    "x": 6,
                    "y": 1
                },
            },
            {
                "start_offset": {
                    "x": 6,
                    "y": 0
                },
                "end_offset": {
                    "x": 7,
                    "y": 1
                },
            }]
    }

    const { getByTestId } = render(
        <GridContext.Provider value={{ cellDimension, lineWidth, canvasDimension, size }}>
            <LandmarksLayer landmarksDimensions={landmarksDimensions} />
        </GridContext.Provider>)
    const context = getByTestId("grid-canvas-landmarks").getContext("2d")
    expect(context.__getEvents()).toMatchSnapshot()
})