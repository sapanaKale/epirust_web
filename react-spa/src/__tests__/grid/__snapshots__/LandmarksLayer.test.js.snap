// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render LandmarksLayer component 1`] = `
Array [
  Object {
    "props": Object {
      "value": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "value": "#ccc",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "height": 6,
      "width": 6,
      "x": 0.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
  Object {
    "props": Object {
      "height": 6,
      "width": 6,
      "x": 4.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
  Object {
    "props": Object {
      "value": "#ccc",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "height": 4,
      "width": 4,
      "x": 10.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
  Object {
    "props": Object {
      "height": 4,
      "width": 4,
      "x": 12.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
]
`;
